### PR TITLE
Refactor to splitt out work that can be done in separate thread.

### PR DESCRIPTION
**Pull request by baldersheim**
@geirst and @toregge PR

### CODE REVIEW

**Review by toregge — APPROVED** [review]

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp [hunk 1]
```diff
@@ -8,135 +8,138 @@
...
-DocumentBucketMover::DocumentBucketMover(IMoveOperationLimiter &limiter)
-    : _limiter(limiter),
-      _bucket(),
-      _source(nullptr),
-      _targetSubDbId(0),
-      _handler(nullptr),
-      _bucketDb(nullptr),
-      _bucketDone(true),
+BucketMover::BucketMover(const BucketId &bucket, const MaintenanceDocumentSubDB *source, uint32_t targetSubDbId,
+                         IDocumentMoveHandler &handler,BucketDBOwner &bucketDb) noexcept
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp:48]: Nitpick: Missing space after '_,_'.

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.h [hunk 2]
```diff
@@ -6,48 +6,113 @@
...
+class MoveOperation;
 
-/**
- * Class used to move all documents in a bucket from a source sub database
- * to a target sub database. The actual moving is handled by a given instance
- * of IDocumentMoveHandler.
- */
-class DocumentBucketMover
+namespace bucketdb {
+    /**
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/documentbucketmover.h:20]: Nitpick: Indentation is off.

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp [hunk 3]
```diff
@@ -8,135 +8,138 @@
...
-    _targetSubDbId = targetSubDbId;
-    _handler = &handler;
-    _bucketDb = &bucketDb;
-    _bucketDone = false;
-    _lastGid = GlobalId();
-    _lastGidValid = false;
+std::pair<std::vector<BucketMover::MoveKey>, bool>
+BucketMover::getKeysToMove(size_t maxDocsToMove) const {
+    std::pair<std::vector<BucketMover::MoveKey>, bool> result;
+    Iterator itr = (_lastGidValid ? _source->meta_store()->upperBound(_lastGid)
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp:62]: Consider using auto.

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp [hunk 4]
```diff
@@ -8,135 +8,138 @@
...
+                                  : _source->meta_store()->lowerBound(_bucket));
+    const Iterator end = _source->meta_store()->upperBound(_bucket);
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp:64]: Consider using auto.

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp [hunk 5]
```diff
@@ -8,135 +8,138 @@
...
-    }
-    if (done) {
+    auto [keys, done] = getKeysToMove(maxDocsToMove);
+    auto moveOps = createMoveOperations(keys);
+    bool allOk = keys.size() == moveOps.size();
+    if (done && allOk) {
         setBucketDone();
     }
-    return true;
+    if (moveOps.empty()) return allOk;
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp:111]: Consider using curly braces.

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.h [hunk 6]
```diff
@@ -6,48 +6,113 @@
...
+  * Class used to move all documents in a bucket from a source sub database
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/documentbucketmover.h:21]: This description must be updated. It is currently identical to the description in DocumentBucketMover.

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.h [hunk 7]
```diff
@@ -6,48 +6,113 @@
...
-                      const document::GlobalId &gid,
-                      storage::spi::Timestamp timestamp);
+    void moveDocument(MoveOperationUP moveOp, IDestructorCallbackSP onDone);
+    MoveOperationUP createMoveOperation(const MoveKey & key);
+};
+}
 
-    void setBucketDone();
+/**
+ * Class used to move all documents in a bucket from a source sub database
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/documentbucketmover.h:89]: Ensure that this description and the description for BucketMover are up to date.

**Review by geirst — COMMENTED** [review]
👍

**Pull request merged**